python/tank/path_cache.py

```diff
@@ -1265,3 +1170,142 @@ def get_secondary_entities(self, path):
 
         return matches
     
+
+    def ensure_all_entries_are_in_shotgun(self, log):
+        """
+        Ensures that all the path cache data in this database is also registered in Shotgun.
+        
+        This will go through each entity in the path cache database and check if it exists in 
+        Shotgun. If not, it will be created.
+        
+        No updates will be made to the path cache database.
+        
+        :param log: Std python logger 
+        """
+
+        SG_BATCH_SIZE = 16
+
+        total_records_added = 0
+
+        log.info("")
+        log.info("Downloading current path data from Shotgun...")
+        
+        project_link = {"type": "Project", 
+                        "id": self._tk.pipeline_configuration.get_project_id() }        
+        
+        sg_data = self._tk.shotgun.find(SHOTGUN_ENTITY, 
+                                        [["project", "is", project_link]],
+                                        [SG_PATH_FIELD, SG_ENTITY_TYPE_FIELD, SG_ENTITY_ID_FIELD])
+        log.info(" - Download complete. Got %s records." % len(sg_data))
+
+        # reshuffle these into a dictionary based on path and entity type
+        # this is so we can do fast lookups later
+        sg_existing_data = {}
+        for p in sg_data:
+            local_path = p[SG_PATH_FIELD].get("local_path") # using get() in case key is missing
+            # key dictionary by local path, entity type and entity id. This is so that we can 
+            # handle secondary entities correctly.
+            dict_key = (local_path, p[SG_ENTITY_TYPE_FIELD], p[SG_ENTITY_ID_FIELD])
+            sg_existing_data[dict_key] = p["id"]
+        
+        # free up memory
+        sg_data = None
+                
+        
+        cursor = self._connection.cursor()
+
+        try:
+            # get all records and check each one against shotgun.
+            pc_data = list(cursor.execute("""select pc.rowid,
+                                                    pc.entity_type, 
+                                                    pc.entity_id, 
+                                                    pc.entity_name, 
+                                                    pc.root, 
+                                                    pc.path, 
+                                                    pc.primary_entity 
+                                             from path_cache pc"""))
+        finally:
+            cursor.close()
+        
+        log.info("")
+        log.info("There are %s paths in the local path cache." % len(pc_data))
+        
+        log.info("")
+        log.info("Starting migration.")
+        
+        sg_records = []
+
+        for sql_record in pc_data:
+            log.debug("Processing db record %s..." % str(sql_record))
```

Comment:
Would be good to give some indication of progress and how many items there are left to process